### ISSUE TITLE
BeforeAll/AfterAll?

### ISSUE THREAD

**Issue by modocache** [onsi/ginkgo#70]
> First of all, thanks so much for this library. It really makes testing in Go a joy.

RSpec includes `before(:all)` and `after(:all)` nodes. These nodes are executed once at the beginning or end of an example group, respectively:

``` ruby
require 'rspec'

RSpec.describe 'before and after all nodes' do
  before(:all) { puts "\nbefore(:all)" }
  after(:all) { puts "\nafter(:all)" }
  it 'is an example' { }
  it 'is another example' { }
end
```

When run, `rspec` outputs:

```
before(:all)
..
after(:all)
```

I was surprised that Ginkgo does not include `BeforeAll` or `AfterAll` functions. Is this a conscious omission? Would you consider a pull request which adds these functions?

**Comment by mml** [comment]
I know this is an ancient issue, but I'll mention another way this is an issue.  In kubernetes, we have a rather large end-to-end test suite.  There is a single suite for the entire project, and so the nested Describe/Context blocks often have very specific needs.

As this is an e2e setup, it's often the case that we need to create certain API objects and wait for them to stabilize in the running server.  This is a high-latency operation.  Currently we do this in BeforeEach blocks, which works but it strongly encourages test writers to build gigantic It() blocks because of the state teardown/construction.  In effect, we have the same amount of test assertions "sharing state" as we would with BeforeAll, but the tests are large, harder to read, and they fail with very little granularity.

BeforeAll/AfterAll or equivalently nested suites would be a useful abstraction for this case.

**Comment by and-hom** [comment]
I need BeforeAll/AfterAll too. I created integration test for my application. It prepares environment, starts app and performs tests. So I'd like to test my app with different config files and different environment variables. And I have more then one test per each configuration.
Now I have to create a package per each test configuration and one more package for common test utils. It's a workaround and looks ugly.
It would be great to create own setUp/tearDown for Context

**Comment by onsi** [comment]
@and-hom can you point me at a concrete or specific example?

PSA:

If you want `BeforeAll` or `AfterAll` you can always add a guard to a `BeforeEach` and `AfterEach` block to ensure it only runs once.

This will not work with paralellized tests or `--randomizeAllSpecs`.  This is one reason why I do not intend to add `BeforeAll` or `AfterAll`.

**Comment by zeekay** [comment]
I think this would be a nice addition. It helps discourage massive `It` blocks, which are a natural response to `BeforeEach` blocks with high-latency setup work (as mentioned by @mml). As per @onsi's suggestion above we implement our own BeforeAll helper like so:

```go
func BeforeAll(fn func()) {
	first := true
	Before(func() {
		if first {
			fn()
			first = false
		}
	})
}
```

Which works well with our parallelized (and randomized) test setup.

This has been a boon in dealing with some particularly unreliable testing infrastructure (rapid iteration of `It` blocks causing test server to explode). Judicious use is of course advised, `BeforeEach` is what you want in most cases...

**Comment by b0o** [comment]
@onsi 
> If you want `BeforeAll` or `AfterAll` you can always add a guard to a BeforeEach and AfterEach block to ensure it only runs once.

I can see how this makes `BeforeAll` possible, but how does it help with emulating `AfterAll`? Adding a guard to an `AfterEach` would just allow you to run it once after the first `It`, right?

I really need an `AfterAll` to reconstruct some high-cost objects.

I've been racking my brain trying to come up with various workarounds, but since the `AfterEach` will be the final thing to run within a `Describe` or `Context`, I don't know how to know that it's really the last one (or if it's even possible to know).

**Comment by srbry** [comment]
@b0o since each describe/ context/ it are really just functions, could you not just use a `defer` on the function to ensure it runs whenever the function finishes? Not actually tested it myself.

**Comment by akutz** [comment]
Hi,

I also would like to see `BeforeAll` implemented for the same reason. Sometimes standing up an environment is expensive.

**Comment by glyoko** [comment]
+1

**Comment by tzvatot** [comment]
> I know this is an ancient issue, but I'll mention another way this is an issue. In kubernetes, we have a rather large end-to-end test suite. There is a single suite for the entire project, and so the nested Describe/Context blocks often have very specific needs.
> 
> As this is an e2e setup, it's often the case that we need to create certain API objects and wait for them to stabilize in the running server. This is a high-latency operation. Currently we do this in BeforeEach blocks, which works but it strongly encourages test writers to build gigantic It() blocks because of the state teardown/construction. In effect, we have the same amount of test assertions "sharing state" as we would with BeforeAll, but the tests are large, harder to read, and they fail with very little granularity.
> 
> BeforeAll/AfterAll or equivalently nested suites would be a useful abstraction for this case.

I'm facing the same issue: working with openshift (kubernetes based) I have a test suite devided to several contexts and specs. Some tests needs a setup before their execution. The setup takes a long time, but it's the same setup for all `It` in the context, which the outcome is a state which the test requires in order to run. Running the setup before each test cause the test duration to exceed the timeout. Besides it consumes a lot of redundant resources.
Having a `beforeAll` and `afterAll` would allow me to setup the state, pass that to each `It` as a context variable and tear it down on at the end of all of them.

**Comment by williammartin** [comment]
@tzvatot Thanks for your feedback! Is this code somewhere I can see?

**Comment by tzvatot** [comment]
> @tzvatot Thanks for your feedback! Is this code somewhere I can see?

Unfortunately no. But we might open it up sometime in the future.

**Comment by wookasz** [comment]
+1 on the `BeforeAll` and `AfterAll`. We have some integration tests which involve indexing to elasticsearch. It can take over a second for the document to be updated in ES in the test environment. We want to be able to trigger the document update and test several assertions. Multiple assertions in a single `It` are not ideal because it's unclear which assertion failed and when a single assertion fails, the remaining ones are not executed.

**Comment by arxeiss** [comment]
Doesn't `BeforeSuite()` and `AfterSuite()` solving exactly this problem? http://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite

> ## Global Setup and Teardown: BeforeSuite and AfterSuite
> Sometimes you want to run some set up code once before the entire test suite and some clean up code once after the entire test suite. For example, perhaps you need to spin up and tear down an external database.
> 
> Ginkgo provides BeforeSuite and AfterSuite to accomplish this. You typically define these at the top-level in the bootstrap file. For example, say you need to set up an external database:

**Comment by tzvatot** [comment]
@arxeiss no, it does not solve the problem. As mentioned in the documentation, the BeforeSuite and AfterSuite are global.

For example, assume I have a single test suite called "regression". This test suite contains several test cases. Few of the tests share the same exact setup, which is costly both in time and resources. So instead of forcing each of those tests to do it's own setup, we like to run that setup once, to create some basic resources that will be used by all those tests, and once they finish, we can remove those resources.

Of course, you may argue that those should be in a separated Suite, but since we have several such "groups" of tests that should share the setup and teardown in the "regression" suite, it means that every real life suite will actually be a group of suites (which opens another question about missing BeforeEachSuite and AfterEachSuite).

**Comment by thallgren** [comment]
The lack of `BeforeAll()` and `AfterAll()` forces any tests that must have common environment prerequisites to live in a package of its own because the `BeforeSuite()` and `AfterSuite()` are package global. This is an unnecessary and annoying limitation in an otherwise excellent test framework.

I have plenty of examples of where I want a `Context()` to have its own environment setup. It can be setting up and tearing down a database, database tables, or table content, modifying a kubernetes cluster, docker images, a daemon process, etc. The time to set up and tear down such environments ranges from a second to a minute. Moving every such context into a suite (and hence a directory) of its own is not an option because the contexts often reflect different aspects of one and the same test subject, or its undesirable to export the functions that are tested.

At present, the global `BeforeSuite()` and `AfterSuite()` limits randomizing specs to one package at a time. That's nice of course, but why can't that limit just as well be on a `Context()` if the user so desires? In my case it would vouch for a much better organization of the specs.

I see that implementations start to pop up that provides `BeforeAll()` and `AfterAll()` such as this [Ciliums e2e testing](https://docs.cilium.io/en/v1.8/contributing/testing/e2e/#ginkgo-extensions). Wouldn't it be nicer to have ginkgo provide this?

**Comment by liuxh0** [comment]
Generally, it does make sense without BeforeAll and AfterAll.
However, while I was writing integration tests recently, I do need them to reduce calling external services, which are time-consuming.

-> Therefore, +1 for BeforeAll / AfterAll. Also would like to know the best practices for integration tests using Ginkgo.

**Comment by blgm** [comment]
Ginkgo 2.0 (see #711) will hopefully add some additional controls to help with this.

**Comment by hamadise** [comment]
We do use Ginkgo to develop a framework to test applications of Openshift/Kubernetes and having BeforeAll is really sweet option. Otherwise, we're forced to do that ourselves (not in a clean way)
In Openshift, we do need to setup the whole application with all the sessions to different Pods/Nodes. 
Then we do run different suites and differents tests. "It" blocks are fully separated and can run in any given order. We can add the whole logic of setting up the app and sessions inside beforeEach. But it takes a significant amount of time. 
Reading the comments, people such as  (@tzvatot @mml ) are also facing similar issue in k8s/openshift.
our code is open and anyone is welcome to check it and see how ginkgo is really useful https://github.com/test-network-function/test-network-function
We're waiting for official release of Ginkgo v2 :)

**Comment by tzvatot** [comment]
@onsi how can I find the PR for this? what version will include it?

**Comment by onsi** [comment]
hey - this shipped in 2.0 and is documented [here](https://onsi.github.io/ginkgo/#ordered-containers).

**Comment by tzvatot** [comment]
@onsi it seems that the BeforeAll/AfterAll is strongly tight with Ordered context. This cause the following issue:
In case of using a DescribeTable with multiple entries, the first entry fails the rest of the entries, because it's `Ordered`. Removing the `Ordered` is not an option as it prevents using `BeforeAll/AfterAll`. 
We need an ability to have BeforeAll/After all, but still be able to have DescribeTable run through all of the entries, without skipping them as soon as the first fails.

**Comment by onsi** [comment]
hey @tzvatot do you mind opening a new issue so we can discuss it there?  i'd like to learn more about what you're doing in the table test to figure out what potential solutions for you are.

**Comment by tzvatot** [comment]
> hey @tzvatot do you mind opening a new issue so we can discuss it there? i'd like to learn more about what you're doing in the table test to figure out what potential solutions for you are.

https://github.com/onsi/ginkgo/issues/1103 submitted

**Comment by onsi** [comment]
cross-posting here since this is the issue that folks thinking about `BeforeAll`/`AfterAll` usually land on:

2.7.0 introduces the `ContinueOnFailure` decorator which resolves #1103 and allows specs in ordered containers to continue running if an earlier spec fails.
